Feature/feedback tab

xurxodev opened the pull request:
### :pushpin: References
* **Issue:** close #31     
* **Related pull-requests:** https://github.com/EyeSeeTea/dhis2-android-sdk/pull/418

###   :gear: branches 
**app**: 
       Origin: :feature/feedback_tab Target: develop-psi
**dhis2-android-SDK**: 
       Origin: feature/download_data_element_and_program_attribute_value_relation
**dhis2-rule-engine**: 
       Origin: android_2.0.1
### :tophat: What is the goal?

The goal is to create a new feedback tab according to this designs  https://www.figma.com/file/XvloHoefcLUTB5oZg8qDgV/RDQA?node-id=20%3A0

### :memo: How is it being implemented?

- I have created functions by flavor to create tabs because feedback tab only should be visible for psi flavor.
- I have created all feedback tab content under psi flavor
- I have used Clean Architecture and MVP for feedback 
- I have used coroutines to asynchronous work
- I have used the current TeiDataRepository to retrieve events but I have created a new ValuesRepository to retrieve values  
- I have created models for feedback rows 
- I have created an immutable tree data structure, a TreeAdapter 
- I have created unit tests for GetFeedback use case
- OnlyFailed and Critical filters are applied over the final feedback tree following the next: a parent is shown if contains a descendant that should be visible by the filters. 

### :boom: How can it be tested?

**Use case 1:** - Feedback for RDQA by indicator should show correctly the feedback
**Use case 2:** - Feedback for RDQA by indicator should show correctly the feedback if the only failed filter is applied
**Use case 3:** - Feedback for RDQA by technical area should show correctly the feedback
**Use case 4:** - Feedback for RDQA by technical area  should show correctly the feedback if the only failed filter is applied
**Use case 5:** - Feedback for HNQIS by All should show correctly the feedback
**Use case 6:** - Feedback for HNQIS by All should show correctly the feedback if the only failed filter is applied
**Use case 7:** - Feedback for HNQIS by Critical should show correctly the feedback
**Use case 8:** - Feedback for HNQIS by Critical should show correctly the feedback if the only failed filter is applied
**Use case 9:** - Feedback for HNQIS by Non-Critical should show correctly the feedback
**Use case 10:** - Feedback for HNQIS by Non-Critical should show correctly the feedback if the only failed filter is applied

### :floppy_disk: Requires DB migration?

- [x] Nope, we can just merge this branch.
- [ ] Yes, but we need to apply it before merging this branch.
- [ ] Yes, it's already applied.

### :art: UI changes?

- [ ] Nope, the UI remains as beautiful as it was before!
- [x] Yeap, here you have some screenshots-

![device-2020-09-23-135924](https://user-images.githubusercontent.com/5593590/94009751-2e832800-fda5-11ea-9d76-5b0083bc3d5d.png)

![device-2020-09-23-140024](https://user-images.githubusercontent.com/5593590/94009792-3a6eea00-fda5-11ea-864e-013694221f5b.png)

![device-2020-09-23-134511](https://user-images.githubusercontent.com/5593590/94008333-1f9b7600-fda3-11ea-83e5-ee9e9238ba0c.png)

ifoche commented on app/src/testPsi/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureRepositoryFunctionsTest.kt:
```diff
@@ -1,4 +1,4 @@
-package org.dhis2.usecases.eventsWithoutRegistration.eventCapture
+package org.dhis2.usescases.eventsWithoutRegistration.eventCapture
```
why in plural "usescases"? I would have said that the plural of usecase is usecases...isn't it?

xurxodev replied:
It's to follow the same approach that main directory:
package org.dhis2.usescases

ifoche reviewed (changes requested):
@xurxodev thanks for the code. I've added a minor comment inline regarding a subpackage name, but apart from that, I've found the following functional "bugs", "info" or "layout" things to mention. To be discussed tomorrow in our meeting:

**Bugs:**
- Feedback doesn't seem to be translated. The name of the DE that appears translated in the form, shows untranslated in the feedback
- Last question in HNQIS feedback appears both in Critical & non-critical questions. There are 2 questions with exactly the same text in the server, but only one has feedback and is introduced in the assessment program. It's confusing because under "All" section you don't see that question twice, so All shows 5 questions and the sum of critical + non-critical is 6...which is not evident to understand. I think both should appear in All if there are 2 with identical text and only once if there's only one that's added to the feedback. If the problem comes from the fact that a critical question is child of a non-critical question, we might take the same decision as with the failed filter, and not display non-critical questions in critical tab (as if it were a filter)
- HNQIS program feedback is showing all hierarchy  when "failed only" filter is activated, if one of the children is failed, but RDQA program is only showing the children. We should harmonize that

**Info:**
- In the current configuration there's no 1.1, when checking that's not evident. We might need to add one so they don't report it
- Legends are also applied in data entry form. That provides the user an idea of whether or not he's doing well...which is probably something that should be reserved to the feedback. We might ask this question to the client

**Layout:**
- Padding for parent-child is not enough in phones (apparently ok in tablets). I think we need to find a solution because it's already quite confusing when filter is applied (maybe adding a little symbol to represent each padding block?)
- Feedback arrow shows also in short feedbacks. This one you were already aware. I don't think we have to implement for this release

xurxodev commented:
@ifoche I have applied the bug fixing:
- I have renamed feedback order in hierarchy DE to follow natural order -> 1-> 1.1 -> 1.1.1
- I have assigned padding to children by level calculating pixels according to the density of pixels (12 dp) 
- I have used displayFormName to show DE names in feedback translated
- I have fixed design problems in mobile devices

ifoche reviewed (approved):
thanks @xurxodev
